### ISSUE TITLE
Add simple publish (generate public view-only url) option

### ISSUE THREAD

**Issue by heyakyra** [fiduswriter/fiduswriter#859]
This combined with #383 would make FidusWriter beat https://github.com/writeas/draftas to the punch. 

StandardNotes, HackMD, etc. all allow you to "publish" a document, which generates a nicely formatted HTML page which you can link to.

**Comment by johanneswilm** [comment]
Hey, @heyakyra that seems to be a popular idea which is why it comes up every now and then. We have discussed whether we should do that in the past. The main reason it doesn't exist is time constraints, but there is also the issue  of hosting documents and keeping stable URLs requiring a somewhat different approach in that one suddenly has to consider situation sin which thousands of people connect to the document. Even if they can only read it, they'll still need to be connected to the websocket channel if they are to receive realtime updates of the document as it is being changed.

**Comment by heyakyra** [comment]
Hi @johanneswilm, thanks for the reply! I don't believe the other's that offer a publish button have the link live update for people who are already viewing it (I think this would also be surprising to readers). It seems very reasonable to me to require a refresh for updates. 

I do understand how anonymous editing creates a much bigger difficulty though! I'm no expert but there have been some nice writeups from others who have taken on the challenge: 
* https://hackernoon.com/building-conclave-a-decentralized-real-time-collaborative-text-editor-a6ab438fe79f
* https://medium.com/@srijancse/how-real-time-collaborative-editing-work-operational-transformation-ac4902d75682

that's not to say, "hey look others did it, now your turn!" just some interesting reading/references that may be useful sometime down the line when resources are available to put to it

**Comment by johanneswilm** [comment]
> that's not to say, "hey look others did it, now your turn!" just some interesting reading/references that may be useful sometime down the line when resources are available to put to it

Understood. And thanks for the links. 

Just a little note on some of that: Creating a prototype collaborative editor is a fun project as the first article says. Making it have all the bells and whistles that one actually requires from an online word processor takes a lot longer. Fidus Writer has been developed since 2012 - a lot by myself but not only me. And still there are things it doesn't have. Something like the kind of realtime collaborative editing described in the articles it first obtained in 2013. Then a second, more stable implementation building upon the ProseMirror engine was launched in summer 2016. That was Fidus Writer 3.0. Right now we are developing version 3.7 and there have been tons of additions since then, but some of the earliest tickets have still not been resolved.

I don't actually think anonymous editing is much different than having a user edit it in terms of programming, except the part about suddenly possibly having to deal with tens of thousands of users in the same document. If a developer were to take it on, I'd think they should be able to do it within max a few weeks.

**Comment by heyakyra** [comment]
> Even if they can only read it, they'll still need to be connected to the websocket channel if they are to receive realtime updates of the document as it is being changed.

I think that this is actually not the expected behavior of a "published" document. Would it be simpler to implement if users had to refresh to see changes?

**Comment by johanneswilm** [comment]
@heyakyra I've thought about this recently as also others have expressed an interest in something like this. I actually think it's two different things - 

A) Some want to be able to give access to others for some quick editing or reading and they always want them to have the newest available document. For example, they may have "finished" writing a specific doc, then emailed their colleague to please take a look at it, but before their colleague gets to it they find some errors that they fix while waiting.

B) Others will want to be able click a button "publish" which then takes a snapshot of the document as it currently is and makes it available on the web for the general public.

I can see both being useful. If we implement A and/or B, we probably need to implement A in a way so that it cannot be used for the purpose of B because A puts a lot more strain on the server. For example, A could be done in a way so that the amount of people connected to the doc is limited to 10. B should be implemented in a way so that it gets to use the same static file caching as other static resources on the site.

**Comment by heyakyra** [comment]
I think an option C would be that the published link always just displays the most recent changes, but doesn't update if the publisher is in the middle of making changes. A and B are both more complex than this, so I think it makes sense to start with the simplest and iterate from there, no?

**Comment by heyakyra** [comment]
StandardNotes publishes to https://listed.to/ and only publishes when you click publish or update, not as you make changes

**Comment by almereyda** [comment]
Yes, publishing (HTML artefacts) to an extra service could offload the read-only visitors from the main service. This could be as simple as a frontend to S3, similar to what people do with https://github.com/lindenlab/caddy-s3-proxy or GitLab/Gitea Pages, but configured automatically.

This is also what https://obsidian-userland.github.io/publish/ does for https://obsidian.md/.

---

While I love the idea A to watch a live-session as a visitor, I believe B is easier to implement, given that production instances already serve static files through a separate HTTP server.

Could we rework the existing HTML export in a way, that generated artefacts can be placed into `/static/` alongside some special CSS in a deterministic manner, e.g. through strictly linking to versions?

https://github.com/fiduswriter/fiduswriter-gitrepo-export shows us today that offloading exports to a third party is very well desireable. With GitLab-/GitHub-Export, we'd even have a way to publish to the Web, using their CI/CD and Pages offers.

**Comment by johanneswilm** [comment]
The `fiduswriter-website` plugin allow for this now.

**Comment by johanneswilm** [comment]
hey @heyakyra no it's not enabled there. The reason is that if we let user's content be shown directly on the internet, we would become liable for what they write. Which means I would have to look at everything that is published constantly and shut down accounts, unpublish neonazi content, etc. .

**Comment by heyakyra** [comment]
Ah you're not in the USA? Section 230 protects against that here from legal issues, though there are other probably ways abuse can happen that are still problematic (platforms marking the entire domain as spam, etc)
https://www.eff.org/issues/cda230

**Comment by heyakyra** [comment]
I see, that makes a lot of sense. Maybe you'd consider paywalling it since that gives easy leverage, but I understand either way as you know best what workload would be involved. Glad its possible via self-hosting though!
